File: examples/rust/flappy_bird/src/game_manager.rs
use crate::bird::Bird;
use crate::constants::*;
use crate::engine::{Engine, KEY_ESCAPE, KEY_R, KEY_SPACE, MOUSE_LEFT};
use crate::pipe::Pipe;
use crate::score::ScoreCounter;

/// Loaded texture handles (created once at init, reused every frame).
pub struct Textures {
    pub background: u64,
    pub base: u64,
    pub pipe: u64,
    pub bird_frames: [u64; 3],
}

impl Textures {
    pub fn load(engine: &Engine, asset_base: &str) -> Self {
        let background = engine.load_texture(&format!("{asset_base}/sprites/background-day.png"));
        let base = engine.load_texture(&format!("{asset_base}/sprites/base.png"));
        let pipe = engine.load_texture(&format!("{asset_base}/sprites/pipe-green.png"));
        let bird_frames = [
            engine.load_texture(&format!("{asset_base}/sprites/bluebird-downflap.png")),
            engine.load_texture(&format!("{asset_base}/sprites/bluebird-midflap.png")),
            engine.load_texture(&format!("{asset_base}/sprites/bluebird-upflap.png")),
        ];
        Self {
            background,
            base,
            pipe,
            bird_frames,
        }
    }
}

/// Orchestrates game state: bird, pipes, scoring, collision, rendering.
pub struct GameManager {
    bird: Bird,
    pipes: Vec<Pipe>,
    score: ScoreCounter,
    textures: Textures,
    pipe_spawn_timer: f32,
}

impl GameManager {
    pub fn new(engine: &Engine, asset_base: &str) -> Self {
        let textures = Textures::load(engine, asset_base);
        let mut score = ScoreCounter::new();
        score.load_textures(engine, asset_base);
        Self {
            bird: Bird::new(),
            pipes: Vec::new(),
            score,
            textures,
            pipe_spawn_timer: 0.0,
        }
    }

    /// Resets all game state for a fresh round.
    pub fn start(&mut self) {
        self.bird.reset();
        self.pipes.clear();
        self.score.reset();
        self.pipe_spawn_timer = 0.0;
    }

    /// Runs one frame of game logic. Returns `false` when the window
    /// should close (user pressed Escape).
    pub fn update(&mut self, engine: &Engine, dt: f32) -> bool {
        // -- Input --------------------------------------------------------
        if engine.is_key_pressed(KEY_ESCAPE) {
            return false;
        }
        if engine.is_key_pressed(KEY_R) {
            self.start();
            return true;
        }

        let jump = engine.is_key_pressed(KEY_SPACE) || engine.is_mouse_button_pressed(MOUSE_LEFT);

        // -- Update bird --------------------------------------------------
        self.bird.update(dt, jump);

        // Ground / ceiling collision
        if self.bird.y + BIRD_HEIGHT > SCREEN_HEIGHT || self.bird.y < 0.0 {
            self.start();
            return true;
        }

        // -- Update pipes and check collisions ----------------------------
        for pipe in &mut self.pipes {
            pipe.update(dt);
            if pipe.collides_with_bird(self.bird.x, self.bird.y, BIRD_WIDTH, BIRD_HEIGHT) {
                self.start();
                return true;
            }
        }

        // -- Spawn new pipes ----------------------------------------------
        self.pipe_spawn_timer += dt;
        if self.pipe_spawn_timer > PIPE_SPAWN_INTERVAL {
            self.pipe_spawn_timer = 0.0;
            self.pipes.push(Pipe::new());
        }

        // -- Remove off-screen pipes and increment score ------------------
        let before = self.pipes.len();
        self.pipes.retain(|pipe| !pipe.is_off_screen());
        let removed = before - self.pipes.len();
        for _ in 0..removed {
            self.score.increment();
        }

        true
    }

    /// Draws all layers in painter's order (back to front).
    pub fn draw(&self, engine: &Engine) {
        let tex = &self.textures;

        // Layer 0: Background
        engine.draw_sprite(
            tex.background,
            BACKGROUND_WIDTH / 2.0,
            BACKGROUND_HEIGHT / 2.0,
            BACKGROUND_WIDTH,
            BACKGROUND_HEIGHT,
            0.0,
        );

        // Layer 1: Score (behind pipes, in front of background)
        self.score.draw(engine);

        // Layer 2: Pipes
        for pipe in &self.pipes {
            // Top pipe (rotated 180 degrees)
            engine.draw_sprite(
                tex.pipe,
                pipe.x + PIPE_IMAGE_WIDTH / 2.0,
                pipe.top_pipe_y() + PIPE_IMAGE_HEIGHT / 2.0,
                PIPE_IMAGE_WIDTH,
                PIPE_IMAGE_HEIGHT,
                std::f32::consts::PI,
            );
            // Bottom pipe (no rotation)
            engine.draw_sprite(
                tex.pipe,
                pipe.x + PIPE_IMAGE_WIDTH / 2.0,
                pipe.bottom_pipe_y() + PIPE_IMAGE_HEIGHT / 2.0,
                PIPE_IMAGE_WIDTH,
                PIPE_IMAGE_HEIGHT,
                0.0,
            );
        }

        // Layer 3: Bird
        let bird_tex = tex.bird_frames[self.bird.animator.frame_index];
        let rotation_rad = self.bird.rotation_deg().to_radians();
        engine.draw_sprite(
            bird_tex,
            self.bird.x + BIRD_WIDTH / 2.0,
            self.bird.y + BIRD_HEIGHT / 2.0,
            BIRD_WIDTH,
            BIRD_HEIGHT,
            rotation_rad,
        );

        // Layer 4: Base / ground (in front of everything)
        engine.draw_sprite(
            tex.base,
            BASE_SPRITE_WIDTH / 2.0,
            SCREEN_HEIGHT + BASE_HEIGHT / 2.0,
            BASE_SPRITE_WIDTH,
            BASE_HEIGHT,
            0.0,
        );
    }
}

File: examples/cpp/flappy_bird/src/game_manager.cpp
#include "game_manager.hpp"

#include <string>

namespace flappy {

// Relative path from the build directory to the shared sprite assets.
// Expected working directory: examples/cpp/flappy_bird/build/
// (run the binary from the build/ subdirectory created by cmake)
static const std::string ASSET_BASE = "../../../csharp/flappy_goud/assets/sprites/";

static goud_texture loadTex(goud::Context& ctx, const std::string& file) {
    goud_texture tex = 0;
    std::string path = ASSET_BASE + file;
    ctx.loadTexture(path.c_str(), tex);
    return tex;
}

void GameManager::init(goud::Context& ctx) {
    backgroundTex_ = loadTex(ctx, "background-day.png");
    baseTex_       = loadTex(ctx, "base.png");
    pipeTex_       = loadTex(ctx, "pipe-green.png");

    birdFrames_[0] = loadTex(ctx, "bluebird-downflap.png");
    birdFrames_[1] = loadTex(ctx, "bluebird-midflap.png");
    birdFrames_[2] = loadTex(ctx, "bluebird-upflap.png");

    for (int i = 0; i < 10; ++i) {
        digitTex_[i] = loadTex(ctx, std::to_string(i) + ".png");
    }

    reset();
}

void GameManager::reset() {
    bird_.reset();
    pipes_.clear();
    score_.reset();
    pipeSpawnTimer_ = 0.0f;
    state_ = GameState::WAITING;
}

bool GameManager::handleInput(goud::Context& ctx) {
    // Escape quits.
    if (ctx.keyDown(static_cast<goud_key>(KEY_ESCAPE))) {
        return false;
    }

    // R restarts.
    if (ctx.keyJustPressed(static_cast<goud_key>(KEY_R))) {
        reset();
    }

    return true;
}

void GameManager::update(goud::Context& ctx, float dt) {
    // Determine jump input.
    bool jumpInput = ctx.keyDown(static_cast<goud_key>(KEY_SPACE))
                  || ctx.mouseDown(static_cast<goud_mouse_button>(MOUSE_BUTTON_LEFT));

    if (state_ == GameState::WAITING) {
        if (jumpInput) {
            state_ = GameState::PLAYING;
        }
        return;
    }

    if (state_ == GameState::GAME_OVER) {
        return;
    }

    // -- PLAYING state --------------------------------------------------------

    // Update bird (physics + animation + jump intent).
    bird_.update(dt, jumpInput);

    // Ground / ceiling collision.
    if (bird_.y() + BIRD_HEIGHT > SCREEN_HEIGHT || bird_.y() < 0.0f) {
        reset();
        return;
    }

    // Update pipes.
    for (auto& pipe : pipes_) {
        if (checkCollision(pipe, bird_.getBounds())) {
            reset();
            return;
        }
    }

    // Move pipes and score for removed (off-screen) ones.
    int removed = updatePipes(pipes_, dt);
    for (int i = 0; i < removed; ++i) {
        score_.increment();
    }

    // Spawn new pipes.
    pipeSpawnTimer_ += dt;
    if (pipeSpawnTimer_ > PIPE_SPAWN_INTERVAL) {
        pipeSpawnTimer_ = 0.0f;
        pipes_.push_back(createPipe());
    }
}

void GameManager::draw(goud::Context& ctx) {
    goud_color white{ 1.0f, 1.0f, 1.0f, 1.0f };

    // Layer 0: Background
    ctx.drawSprite(
        backgroundTex_,
        BACKGROUND_WIDTH / 2.0f,
        BACKGROUND_HEIGHT / 2.0f,
        BACKGROUND_WIDTH,
        BACKGROUND_HEIGHT,
        0.0f,
        white
    );

    // Layer 1: Score (behind pipes, in front of background)
    score_.draw(ctx, digitTex_);

    // Layer 2: Pipes
    for (const auto& pipe : pipes_) {
        drawPipe(ctx, pipe, pipeTex_);
    }

    // Layer 3: Bird
    goud_texture birdTex = birdFrames_[bird_.animFrame()];
    bird_.draw(ctx, birdTex);

    // Layer 4: Base / ground (in front of everything)
    ctx.drawSprite(
        baseTex_,
        BASE_SPRITE_WIDTH / 2.0f,
        SCREEN_HEIGHT + BASE_HEIGHT / 2.0f,
        BASE_SPRITE_WIDTH,
        BASE_HEIGHT,
        0.0f,
        white
    );
}

}  // namespace flappy
